File: packages/authentication/ts/src/hmac-jwt.ts
import { createHmac, timingSafeEqual } from "node:crypto";
import type { AuthContext, AuthResult, Authenticator } from "@botiva/core";

export interface HmacJwtOptions {
    /** Shared HS256 secret the token was signed with. */
    secret: string;
    /** JWT claim carrying the user identity. Default "sub". */
    subjectClaim?: string;
    /** Clock-skew tolerance in seconds when checking `exp`/`nbf`. Default 0. */
    clockToleranceSec?: number;
}

/**
 * Verifies an HS256 JSON Web Token with zero dependencies (`node:crypto`).
 * Checks the signature, `exp` and `nbf`, then maps the subject claim to the
 * verified userId and exposes the full payload as claims.
 *
 *   new HmacJwtAuthenticator({ secret: process.env.JWT_SECRET! })
 */
export class HmacJwtAuthenticator implements Authenticator {
    readonly #secret: string;
    readonly #subject: string;
    readonly #skew: number;

    constructor(opts: HmacJwtOptions) {
        if (!opts?.secret) throw new Error("HmacJwtAuthenticator requires a secret.");
        this.#secret = opts.secret;
        this.#subject = opts.subjectClaim ?? "sub";
        this.#skew = opts.clockToleranceSec ?? 0;
    }

    authenticate(ctx: AuthContext): AuthResult {
        if (!ctx.token) return { ok: false, reason: "missing token" };
        const payload = this.#verify(ctx.token);
        if (!payload) return { ok: false, reason: "invalid or expired token" };
        const userId = payload[this.#subject];
        if (typeof userId !== "string" || userId.length === 0) {
            return { ok: false, reason: `token missing "${this.#subject}" claim` };
        }
        return { ok: true, userId, claims: payload };
    }

    #verify(token: string): Record<string, unknown> | null {
        const parts = token.split(".");
        if (parts.length !== 3) return null;
        const [headerB64, payloadB64, signatureB64] = parts as [string, string, string];

        let expected: Buffer;
        let actual: Buffer;
        try {
            expected = createHmac("sha256", this.#secret).update(`${headerB64}.${payloadB64}`).digest();
            actual = Buffer.from(signatureB64, "base64url");
        } catch {
            return null;
        }
        if (expected.length !== actual.length || !timingSafeEqual(expected, actual)) return null;

        let header: Record<string, unknown>;
        let payload: Record<string, unknown>;
        try {
            header = JSON.parse(Buffer.from(headerB64, "base64url").toString("utf8"));
            payload = JSON.parse(Buffer.from(payloadB64, "base64url").toString("utf8"));
        } catch {
            return null;
        }
        if (header.alg !== "HS256") return null;

        const now = Math.floor(Date.now() / 1000);
        if (typeof payload.exp === "number" && now > payload.exp + this.#skew) return null;
        if (typeof payload.nbf === "number" && now + this.#skew < payload.nbf) return null;
        return payload;
    }
}

File: packages/authentication/go/hmac_jwt.go
package authentication

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/json"
	"strings"
	"time"

	botiva "github.com/aimtune/botiva/core"
)

// HmacJwtAuthenticator verifies an HS256 JSON Web Token with the standard
// library only: it checks the signature, `exp` and `nbf`, then maps the subject
// claim to the verified userId and exposes the full payload as claims.
//
//	authentication.NewHmacJwtAuthenticator(authentication.HmacJwtOptions{Secret: secret})
type HmacJwtAuthenticator struct {
	secret  []byte
	subject string
	skewSec int64
}

// HmacJwtOptions configure the JWT authenticator.
type HmacJwtOptions struct {
	Secret            string // shared HS256 secret
	SubjectClaim      string // claim carrying the userId; default "sub"
	ClockToleranceSec int64  // skew tolerance for exp/nbf; default 0
}

// NewHmacJwtAuthenticator panics if Secret is empty.
func NewHmacJwtAuthenticator(opts HmacJwtOptions) *HmacJwtAuthenticator {
	if opts.Secret == "" {
		panic("authentication: HmacJwtAuthenticator requires a secret")
	}
	subject := opts.SubjectClaim
	if subject == "" {
		subject = "sub"
	}
	return &HmacJwtAuthenticator{secret: []byte(opts.Secret), subject: subject, skewSec: opts.ClockToleranceSec}
}

// Authenticate verifies the token in ac.Token.
func (a *HmacJwtAuthenticator) Authenticate(_ context.Context, ac botiva.AuthContext) (botiva.AuthResult, error) {
	if ac.Token == "" {
		return botiva.AuthResult{OK: false, Reason: "missing token"}, nil
	}
	payload := a.verify(ac.Token)
	if payload == nil {
		return botiva.AuthResult{OK: false, Reason: "invalid or expired token"}, nil
	}
	userID, ok := payload[a.subject].(string)
	if !ok || userID == "" {
		return botiva.AuthResult{OK: false, Reason: "token missing \"" + a.subject + "\" claim"}, nil
	}
	return botiva.AuthResult{OK: true, UserID: userID, Claims: payload}, nil
}

func (a *HmacJwtAuthenticator) verify(token string) map[string]any {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil
	}
	signingInput := parts[0] + "." + parts[1]
	sig, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return nil
	}
	mac := hmac.New(sha256.New, a.secret)
	mac.Write([]byte(signingInput))
	if subtle.ConstantTimeCompare(sig, mac.Sum(nil)) != 1 {
		return nil
	}

	headerRaw, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil
	}
	var header map[string]any
	if json.Unmarshal(headerRaw, &header) != nil || header["alg"] != "HS256" {
		return nil
	}
	payloadRaw, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil
	}
	var payload map[string]any
	if json.Unmarshal(payloadRaw, &payload) != nil {
		return nil
	}
	now := time.Now().Unix()
	if exp, ok := payload["exp"].(float64); ok && now > int64(exp)+a.skewSec {
		return nil
	}
	if nbf, ok := payload["nbf"].(float64); ok && now+a.skewSec < int64(nbf) {
		return nil
	}
	return payload
}
